java/org/apache/tomcat/util/IntrospectionUtils.java

```diff
@@ -285,10 +285,19 @@ public static Object getProperty(Object o, String name) {
     public static String replaceProperties(String value,
             Hashtable<Object,Object> staticProp, PropertySource dynamicProp[],
             ClassLoader classLoader) {
+            return replaceProperties(value, staticProp, dynamicProp, classLoader, 0);
+    }
 
+    private static String replaceProperties(String value,
+            Hashtable<Object,Object> staticProp, PropertySource dynamicProp[],
+            ClassLoader classLoader, int iterationCount) {
         if (value.indexOf('$') < 0) {
```

Comment:
We should be looking for `${` and not `$`. Class names often contain `$` because of nested classes, and class names are often used in system properties. Plus, if we are looking for `${` we should look for `${`.

I realize that this is existing code, but we may as well improve it while we're in here.